File: typespec/admin/authorization/management.go
package authorization

import adminspec "github.com/vetchium/src/typespec/admin"

type SetPermissionsRequest struct {
	AdminUserID adminspec.AdminUserID `json:"admin_user_id"`
	Permissions []AdminPermissionID   `json:"permissions"`
}

func (r SetPermissionsRequest) Validate() []string {
	fields := make([]string, 0, 2)
	if !adminspec.IsAdminUserID(r.AdminUserID) {
		fields = append(fields, "admin_user_id")
	}
	if !ValidatePermissions(r.Permissions) {
		fields = append(fields, "permissions")
	}
	return fields
}

File: typespec/admin/authorization/management.ts
import { type AdminUserID, isAdminUserID } from "../types.ts";
import { type AdminPermissionID, validatePermissions } from "./types.ts";

export interface SetPermissionsRequest {
  admin_user_id: AdminUserID;
  permissions: AdminPermissionID[];
}

export function validateSetPermissionsRequest(
  request: SetPermissionsRequest,
): string[] {
  const fields: string[] = [];
  if (!isAdminUserID(request.admin_user_id)) {
    fields.push("admin_user_id");
  }
  if (!validatePermissions(request.permissions)) {
    fields.push("permissions");
  }
  return fields;
}
